### DISCUSSION TITLE
sparkfun NanoBeacon board in100 -> BLE_gateway -> ble_monitor parsing data

### DISCUSSION THREAD

**Discussion by maromme** [custom-components/ble_monitor#1169] [General]
Hi,
I got myselfe some [sparkfun NanoBeacon board in100](https://www.sparkfun.com/products/21327 ) for retrofitting wallswitches. The Beacon is sending out it's GPIO status as a hex numer using the "Manufacturer Specific Data". For that part everything works fine.

Then I took an esp32 and set it up das an BLE gateway following [that](https://community.home-assistant.io/t/diy-homeassistant-passive-ble-gateway/551636) description. That works fine as well.

As I understand that part of the code written to the esp32 is sending the advertised date to HA by triggering an event in HA:
```
ble_gateway:
  id: blegateway
  on_ble_advertise:
    then:
      homeassistant.event:
        event: esphome.on_ble_advertise
        data:
          packet: !lambda return packet;
```
In HA I install the ble_monitor integration and add that automation:
```
automation:
  - alias: ESPHome BLE Advertise
    mode: queued
    trigger:
      - platform: event
        event_type: esphome.on_ble_advertise
    action:
      - service: ble_monitor.parse_data
        data:
          packet: "{{ trigger.event.data.packet }}"
```

Now the BLE_monitor should be parsing the the BLE advertising  - as far as I understood. But my beacon is not within the list of supported devices for the ble_monitor, I thougt of parsing the data myselfe. But are a bit lost there and don't really know wicht approach I should follow:
- try customizing the parser of the ble_monitor (is the any simple example?)
- or rather don't go for the ble_monitor at all and do parsing straight in the auomation as the data is quite simple
`0x<GPIO 1byte 0 0><VCC 1byte 0 0>`
- or is there another approach? I read about the native HA integration for BLE

Thanks a lot,
Markus

**Comment by Ernst79** [comment]
The part that is missing is a parser that translate the data to something HA can understand. In the folder `ble_parser`, you can see all the parsers that are available right now. One option is to add a parser for your device, and add the manufacturer id in `init.py`, such that it gets recognized. 

However, a smarter approach imho would be to use [BTHome](https://bthome.io) format for your data. BTHome is an open format that you can use to forward your data, Data send in this format should be automatically recognized with a corresponding (binary) sensor entity or event. 

When using BTHome, you have two options
1. use the native HA integration (BTHome). You have to use a (Bluetooth BLE Proxy](https://esphome.io/components/bluetooth_proxy.html) to forward data in stead of BLE gateway if you have to extend the range. 
2. use BLE monitor. In that case, you can use BLE gateway like you did now. 

Regarding BTHome, you could use the button event (0x3A) to send button clicks (events) to HA. Or you could use a binary sensor (0x0F)

Note that BLE monitor does not support events. I would go for the native HA integration therefore.

**Reply by maromme** [reply]
Thanks for your reply. Sounds like I should investigate more into BThome. If I get the in100 talking BThome and an simple BLE USB sticks range is enough I wouldn't need any additional hardware and custom integrations. 

About that in100 I can't find much more than the manufacturer information. The internet is not helping much on the in100 in combination with BThome. And as the in100 is "programming" free configuration I've got no clue how to set it up.

**Reply by maromme** [reply]
If someone more capable than me, wants a nanobeacon for testing I wouldn't mind sending one ;-) Just expecting an answer if it's possible to set it up for BThome

**Reply by Ernst79** [reply]
If I understand the docs https://cdn.sparkfun.com/assets/b/e/d/a/f/NanoBeacon_Config_Tool_User_Guide_EN.pdf correctly, you can set a custom advertisement (user defined data), see section 2.1.2.1. It’s not clear to me if you can send data as service data (0x16), instead of manufacturer specific data (0xFF). That is the big question you will have to find out first.

**Reply by Ernst79** [reply]
I’m not sure the device can send service data (0x16). It seems it can only send manufacturer specific data, and BTHome is using the service data (0xFF)

**Reply by Ernst79** [reply]
We can also go for the option that we create a parser for manufacturer specific data, which has 0x0505 as default. Little issue is that the custom format is custom, so we are not able to tell the order of data. But we could define a combination of a device name that matches with a data. As long as we clearly define it in the docs, we should be good.

**Reply by Ernst79** [reply]
Could you post a raw data example. Perhaps the custom format has a byte that tells us what type of data is what.

**Reply by maromme** [reply]
hey thanks for your help! I got the "user defined data" working. I can create raw data like `04-16-05-05-C1` The first 4 bytes are manually set, the last "c1" is the assigned GPIO status.

That's what it looks like in the config tool:
![grafik](https://github.com/custom-components/ble_monitor/assets/60754017/00b49d4a-f6eb-4bec-9e73-7362a58b35d0)
or 
![grafik](https://github.com/custom-components/ble_monitor/assets/60754017/62e2cb92-7b1f-4337-b61a-9695bcb53588)

The tool might be more flexible than thought.

**Reply by Ernst79** [reply]
Ah, that looks promising. Now, replace the 0505 with D2FC (uuid of BTHome), followed by 40 device info byte, and than e.g. 0F for a generic binary sensor.

so, you will need 0616D2FC400Fxx, where xx is 01 for True , 00 for False

**Reply by Ernst79** [reply]
Hmmm, the C1 is possibly an issue, a binary sensor is either 00 or 01. You can also use, as a first test, the count sensor 0616D2FC4009C1 will give you a count sensor in HA with value 193. I’m not sure where C1 stands for in your case. 

I would use the official BTHome integration in HA btw, not BLE monitor. Both should work, but the built in integration supports more (read all) BTHome sensor types.

**Reply by maromme** [reply]
now i start understanding the BThome syntax.

the gpio status is the status of all GPIOs of the beacon as a hex number. that's what c1 stands for.

as the in100 offers 3 advertising sets individually triggert. that way I can easily use it for having 3 binary sensors. I just have to hard code the raw data. but maybe I find some more I'm the inplay docs. and also I've got more reading on the BThome docs. If I've got more I come back. might be interesting for others as well...

**Reply by maromme** [reply]
I got it working. Without installing "bluetooth" oder "BThome" integration in HomeAssistant. Nice thing. Just a small remaining issue: It's very slow. Way to slow for a push button. I ordered a supported Bluetooth dongle and will test it without the bluetooth-proxy. 

configuration for esp home:
```
esphome:
  name: ble-proxy-1
  friendly_name: BLE Proxy 1
  platformio_options: 
    board_build.flash_mode: dio 
    
esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    sdkconfig_options: 
      CONFIG_BT_BLE_42_FEATURES_SUPPORTED: y 
      CONFIG_BT_BLE_50_FEATURES_SUPPORTED: y
      CONFIG_ESP_TASK_WDT_TIMEOUT_S: "10" 
      variant: ESP32C3

# WIFI SETTINGS ETC.

esp32_ble_tracker:
  scan_parameters:
    active: true

bluetooth_proxy:
  active: true
```
configuration of the in100:
![grafik](https://github.com/custom-components/ble_monitor/assets/60754017/2fc17e66-3b1d-4744-827d-3b0c1db42435)

result in HomeAssitant:
![grafik](https://github.com/custom-components/ble_monitor/assets/60754017/7199db8d-0ca0-4302-b673-62d98b190cd2)

**Reply by maromme** [reply]
amazon been really fast, just got that dongle. installed both "Bluetooth" and "BThome" integrations. unplugged the proxy and now it's much faster. still small delays, but that might also be refresh delays of the dashboard. I'll try it with some relayboard later.

**Reply by maromme** [reply]
well, I've to admit, I'm a bit lost. I manage to get data to Home assistant. But it's not reliable. Sometimes it's fast, sometimes the advertising are completely "lost". In the monitor app on my mobile it's running as expected. think reception can't be the problem, as it's just 3 meters to the Bluetooth dongle. I also don't get every triggered advertising in the HA logs.

**Comment by Ernst79** [comment]
It might be interference of WiFi, try to put your dongle on an extension cable as far away from any other wifi or Bluetooth radio and from USB3 devices. These are known to interfere with Bluetooth, as they use the 2.4 GHz band.
